### PR TITLE
Hive read path documentation for HadoopCatalog tables

### PR DESCRIPTION
**Pull request by massdosage**

### CODE REVIEW

File: site/docs/hive.md [hunk 1]
```diff
@@ -84,7 +84,32 @@ You should now be able to issue Hive SQL `SELECT` queries using the above table
...
+##### Create a Hive table
+Now overlay a Hive table on top of this Iceberg table by issuing Hive DDL like so:
+```sql
+CREATE EXTERNAL TABLE table_a 
+STORED BY 'org.apache.iceberg.mr.hive.HiveIcebergStorageHandler' 
+LOCATION 'hdfs://some_bucket/some_path/database_a/table_c';
+```
+
+#### Query the Iceberg table via Hive
+TODO: why does below work if no config settings are set in Hive but fails if we add `set iceberg.mr.catalog=hadoop` like the code suggests we need to do?
```
Comment (massdosage) [inline site/docs/hive.md:102]: @pvary So for the above, if I create the Hive table but don't set `iceberg.mr.catalog=hadoop` or then I can actually query it successfully. However if I *do* set these (which the code at https://github.com/apache/iceberg/blob/master/mr/src/test/java/org/apache/iceberg/mr/hive/TestTables.java#L133 suggests one should) then I can't query the table. Instead I get this stack trace:
```
2020-11-26T16:46:36,531 INFO  [f94f328b-fe24-4cbc-8723-423fb6ffea6e main([])]: mr.Catalogs (Catalogs.java:loadCatalog(212)) - Loaded Hadoop catalog HadoopCatalog{name=hadoop, location=/tmp/iceberg-test-hadoop-tables}
2020-11-26T16:46:36,533 ERROR [f94f328b-fe24-4cbc-8723-423fb6ffea6e main([])]: hive.log (MetaStoreUtils.java:getDeserializer(456)) - error in initSerDe: org.apache.hadoop.hive.serde2.SerDeException Please provide an existing table or a valid schema
org.apache.hadoop.hive.serde2.SerDeException: Please provide an existing table or a valid schema
        at org.apache.iceberg.mr.hive.HiveIcebergSerDe.initialize(HiveIcebergSerDe.java:62) ~[?:?]
        at org.apache.hadoop.hive.serde2.AbstractSerDe.initialize(AbstractSerDe.java:54) ~[hive-exec-2.3.6-amzn-2.jar:2.3.6-amzn-2]
        at org.apache.hadoop.hive.serde2.SerDeUtils.initializeSerDe(SerDeUtils.java:533) ~[hive-exec-2.3.6-amzn-2.jar:2.3.6-amzn-2]
        at org.apache.hadoop.hive.metastore.MetaStoreUtils.getDeserializer(MetaStoreUtils.java:450) ~[hive-exec-2.3.6-amzn-2.jar:2.3.6-amzn-2]
        at org.apache.hadoop.hive.metastore.MetaStoreUtils.getDeserializer(MetaStoreUtils.java:437) ~[hive-exec-2.3.6-amzn-2.jar:2.3.6-amzn-2]
        at org.apache.hadoop.hive.ql.metadata.Table.getDeserializerFromMetaStore(Table.java:281) ~[hive-exec-2.3.6-amzn-2.jar:2.3.6-amzn-2]
        at org.apache.hadoop.hive.ql.metadata.Table.getDeserializer(Table.java:263) ~[hive-exec-2.3.6-amzn-2.jar:2.3.6-amzn-2]
        at org.apache.hadoop.hive.ql.metadata.Hive.createTable(Hive.java:838) ~[hive-exec-2.3.6-amzn-2.jar:2.3.6-amzn-2]
        at org.apache.hadoop.hive.ql.metadata.Hive.createTable(Hive.java:872) ~[hive-exec-2.3.6-amzn-2.jar:2.3.6-amzn-2]
        at org.apache.hadoop.hive.ql.exec.DDLTask.createTable(DDLTask.java:4356) ~[hive-exec-2.3.6-amzn-2.jar:2.3.6-amzn-2]
        at org.apache.hadoop.hive.ql.exec.DDLTask.execute(DDLTask.java:354) ~[hive-exec-2.3.6-amzn-2.jar:2.3.6-amzn-2]
        at org.apache.hadoop.hive.ql.exec.Task.executeTask(Task.java:199) ~[hive-exec-2.3.6-amzn-2.jar:2.3.6-amzn-2]
        at org.apache.hadoop.hive.ql.exec.TaskRunner.runSequential(TaskRunner.java:100) ~[hive-exec-2.3.6-amzn-2.jar:2.3.6-amzn-2]
        at org.apache.hadoop.hive.ql.Driver.launchTask(Driver.java:2183) ~[hive-exec-2.3.6-amzn-2.jar:2.3.6-amzn-2]
        at org.apache.hadoop.hive.ql.Driver.execute(Driver.java:1839) ~[hive-exec-2.3.6-amzn-2.jar:2.3.6-amzn-2]
        at org.apache.hadoop.hive.ql.Driver.runInternal(Driver.java:1526) ~[hive-exec-2.3.6-amzn-2.jar:2.3.6-amzn-2]
        at org.apache.hadoop.hive.ql.Driver.run(Driver.java:1237) ~[hive-exec-2.3.6-amzn-2.jar:2.3.6-amzn-2]
        at org.apache.hadoop.hive.ql.Driver.run(Driver.java:1227) ~[hive-exec-2.3.6-amzn-2.jar:2.3.6-amzn-2]
        at org.apache.hadoop.hive.cli.CliDriver.processLocalCmd(CliDriver.java:233) ~[hive-cli-2.3.6-amzn-2.jar:2.3.6-amzn-2]
        at org.apache.hadoop.hive.cli.CliDriver.processCmd(CliDriver.java:184) ~[hive-cli-2.3.6-amzn-2.jar:2.3.6-amzn-2]
        at org.apache.hadoop.hive.cli.CliDriver.processLine(CliDriver.java:403) ~[hive-cli-2.3.6-amzn-2.jar:2.3.6-amzn-2]
        at org.apache.hadoop.hive.cli.CliDriver.executeDriver(CliDriver.java:821) ~[hive-cli-2.3.6-amzn-2.jar:2.3.6-amzn-2]
        at org.apache.hadoop.hive.cli.CliDriver.run(CliDriver.java:759) ~[hive-cli-2.3.6-amzn-2.jar:2.3.6-amzn-2]
        at org.apache.hadoop.hive.cli.CliDriver.main(CliDriver.java:686) ~[hive-cli-2.3.6-amzn-2.jar:2.3.6-amzn-2]
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_252]
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_252]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_252]
        at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_252]
        at org.apache.hadoop.util.RunJar.run(RunJar.java:239) ~[hadoop-common-2.8.5-amzn-6.jar:?]
        at org.apache.hadoop.util.RunJar.main(RunJar.java:153) ~[hadoop-common-2.8.5-amzn-6.jar:?]
Caused by: org.apache.iceberg.exceptions.NoSuchTableException: Table does not exist: test.iceberg_table_from_hadoop_catalog
        at org.apache.iceberg.BaseMetastoreCatalog.loadTable(BaseMetastoreCatalog.java:108) ~[?:?]
        at org.apache.iceberg.mr.Catalogs.loadTable(Catalogs.java:100) ~[?:?]
        at org.apache.iceberg.mr.Catalogs.loadTable(Catalogs.java:92) ~[?:?]
        at org.apache.iceberg.mr.hive.HiveIcebergSerDe.initialize(HiveIcebergSerDe.java:60) ~[?:?]
        ... 29 more
2020-11-26T16:46:36,533 ERROR [f94f328b-fe24-4cbc-8723-423fb6ffea6e main([])]: exec.DDLTask (DDLTask.java:failed(639)) - org.apache.hadoop.hive.ql.metadata.HiveException: java.lang.RuntimeException: MetaException(message:org.apache.hadoop.hive.serde2.SerDeException Please provide an existing table or a valid schema)
        at org.apache.hadoop.hive.ql.metadata.Hive.createTable(Hive.java:867)
        at org.apache.hadoop.hive.ql.metadata.Hive.createTable(Hive.java:872)
        at org.apache.hadoop.hive.ql.exec.DDLTask.createTable(DDLTask.java:4356)
        at org.apache.hadoop.hive.ql.exec.DDLTask.execute(DDLTask.java:354)
        at org.apache.hadoop.hive.ql.exec.Task.executeTask(Task.java:199)
        at org.apache.hadoop.hive.ql.exec.TaskRunner.runSequential(TaskRunner.java:100)
        at org.apache.hadoop.hive.ql.Driver.launchTask(Driver.java:2183)
        at org.apache.hadoop.hive.ql.Driver.execute(Driver.java:1839)
        at org.apache.hadoop.hive.ql.Driver.runInternal(Driver.java:1526)
        at org.apache.hadoop.hive.ql.Driver.run(Driver.java:1237)
        at org.apache.hadoop.hive.ql.Driver.run(Driver.java:1227)
        at org.apache.hadoop.hive.cli.CliDriver.processLocalCmd(CliDriver.java:233)
        at org.apache.hadoop.hive.cli.CliDriver.processCmd(CliDriver.java:184)
        at org.apache.hadoop.hive.cli.CliDriver.processLine(CliDriver.java:403)
        at org.apache.hadoop.hive.cli.CliDriver.executeDriver(CliDriver.java:821)
        at org.apache.hadoop.hive.cli.CliDriver.run(CliDriver.java:759)
        at org.apache.hadoop.hive.cli.CliDriver.main(CliDriver.java:686)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.apache.hadoop.util.RunJar.run(RunJar.java:239)
        at org.apache.hadoop.util.RunJar.main(RunJar.java:153)
Caused by: java.lang.RuntimeException: MetaException(message:org.apache.hadoop.hive.serde2.SerDeException Please provide an existing table or a valid schema)
        at org.apache.hadoop.hive.ql.metadata.Table.getDeserializerFromMetaStore(Table.java:283)
        at org.apache.hadoop.hive.ql.metadata.Table.getDeserializer(Table.java:263)
        at org.apache.hadoop.hive.ql.metadata.Hive.createTable(Hive.java:838)
        ... 22 more
Caused by: MetaException(message:org.apache.hadoop.hive.serde2.SerDeException Please provide an existing table or a valid schema)
        at org.apache.hadoop.hive.metastore.MetaStoreUtils.getDeserializer(MetaStoreUtils.java:458)
        at org.apache.hadoop.hive.metastore.MetaStoreUtils.getDeserializer(MetaStoreUtils.java:437)
        at org.apache.hadoop.hive.ql.metadata.Table.getDeserializerFromMetaStore(Table.java:281)
        ... 24 more
```
Any ideas what's going on? I've got a feeling I'm missing something obvious, it's been a long day ;) I'll try debug it tomorrow, just thought you might have some ideas :)

File: site/docs/hive.md [hunk 2]
```diff
@@ -84,7 +84,32 @@ You should now be able to issue Hive SQL `SELECT` queries using the above table
...
+```sql
+SELECT * from table_c;
+```
+
 ### Features
 
 #### Predicate pushdown
 Pushdown of the Hive SQL `WHERE` clause has been implemented so that these filters are used at the Iceberg TableScan level as well as by the Parquet and ORC Readers.
+
+#### Column Projection
```
Comment (massdosage) [inline site/docs/hive.md:114]: This isn't related to the HadoopCatalog but since https://github.com/apache/iceberg/pull/1417 was just merged I thought I might as well just put it here.

File: site/docs/hive.md (hunk 1 above)
Comment (shardulm94) [reply site/docs/hive.md:102]: Shouldn't the catalog to use also be specified when creating the table? It seems odd that the consumer has to be aware of the catalog used to store the table. Also what if we need to read multiple tables stored in different underlying catalogs?

Comment (pvary) [reply site/docs/hive.md:102]: Currently we expect that the catalog is set in the global config (for example: hive-site.xml). Every Iceberg table will use that Catalog. I do not see the exact commands Adrian has executed, but I think Shardul is right and the global config does not contain the catalog settings, and the required values were set between the table creation and the query. Adrian, could you please try:
```
set catalog;
create table;
-- insert data from spark
select from table
```

As for Shardul's question: @lcspinter is started a conversation on the dev list about enabling table level catalog configuration.

Thanks, Peter

Comment (massdosage) [reply site/docs/hive.md:102]: Thanks @pvary, so what I am doing this is...

First I create the table using the Spark API something along the lines of this:
```
HadoopCatalog catalog = new HadoopCatalog(conf, warehousePath);
TableIdentifier tableId = TableIdentifier.of("hadoop_catalog_db", "iceberg_table_from_hadoop_catalog");
catalog.createTable(tableId, schema, spec);
...
Dataset<Row> ds = spark.createDataFrame(hotels, Hotel.class);
ds.write().format("iceberg").mode("append").save(table.location());
```
This then creates the Iceberg table at `hdfs://host/tmp/iceberg-test-hadoop-tables/hadoop_catalog_db/iceberg_table_from_hadoop_catalog/` - I can see the data and metadata folder on HDFS. So far so good. Then I go to a Hive shell, add the `iceberg-hive-runtime.jar` to the path and then try create a Hive table on top of the above like so:
```
CREATE EXTERNAL TABLE test.iceberg_table_from_hadoop_catalog STORED BY 'org.apache.iceberg.mr.hive.HiveIcebergStorageHandler' LOCATION 'hdfs://host/tmp/iceberg-test-hadoop-tables/hadoop_catalog_db/iceberg_table_from_hadoop_catalog/';
```
This looks like its created OK, I check by doing:
```
show create table test.iceberg_table_from_hadoop_catalog;
OK
CREATE EXTERNAL TABLE `test.iceberg_table_from_hadoop_catalog`(
  `id` bigint COMMENT 'from deserializer', 
  `name` string COMMENT 'from deserializer')
ROW FORMAT SERDE 
  'org.apache.iceberg.mr.hive.HiveIcebergSerDe' 
STORED BY 
  'org.apache.iceberg.mr.hive.HiveIcebergStorageHandler' 
WITH SERDEPROPERTIES ( 
  'serialization.format'='1')
LOCATION
  'hdfs://host/tmp/iceberg-test-hadoop-tables/hadoop_catalog_db/iceberg_table_from_hadoop_catalog'
TBLPROPERTIES (
  'table_type'='ICEBERG', 
  'transient_lastDdlTime'='1606485921')
```
At this point I can successfully perform SELECT statements on the table without setting any conf properties, which surprised me, it seems that this is going via the "Hadoop Tables" path since there is a `location` on the table and it actually works! Then I tried doing the equivalent of what I did in order to read Hive Catalog tables which is to set the value for `iceberg.mr.catalog`:
```
set iceberg.mr.catalog=hadoop;
```
I would expect this to now activate the "Hadoop Catalog" path and the SELECT statement to succeed but instead it fails with the error above. Can you think of anything I'm doing wrong?

Comment (massdosage) [reply site/docs/hive.md:102]: @shardulm94 I agree with you, at the moment I'm just documenting how it works currently but the idea is for us all to take a step back from the implementation details and look at it through the lens of an end user and make changes. In our earlier implementation we did store the catalog type as a table property so it didn't have to be specified by the end user, I'm not sure how this got lost along the way. I'll test shortly whether this still works or not. I also think we should rename the property to not have `mr` in it. As @pvary says there is some discussion on the mailing list about this and that should lead to us agreeing to make some changes to how this implemented and then updating these docs accordingly.

Comment (pvary) [reply site/docs/hive.md:102]: @massdosage: Would it be a problem with the warehouse path for the HadoopCatalog: "location=/tmp/iceberg-test-hadoop-tables"? My guess is that this should be "hdfs://host/tmp/iceberg-test-hadoop-tables", and so the HadoopCatalog is not able to read the table data.

Comment (massdosage) [reply site/docs/hive.md:102]: OK, I think we can put the `iceberg.mr.catalog` as a table property and this will be used (so it doesn't have to be done at the Hive conf level). I tried this on the HadoopCatalog table above and get the same error when trying to SELECT as when I set the environment variable (which doesn't solve my problem but at least is consistent :) ). @pvary I see the tables automatically created by HiveCatalog don't have that as a table property but perhaps they should if the "hive execution enabled" flag is set to true?

Comment (massdosage) [reply site/docs/hive.md:102]: > @massdosage: Would it be a problem with the warehouse path for the HadoopCatalog: "location=/tmp/iceberg-test-hadoop-tables"? My guess is that this should be "hdfs://host/tmp/iceberg-test-hadoop-tables", and so the HadoopCatalog is not able to read the table data.

OK, this jogged me into realising that I also need to set "iceberg.mr.catalog.hadoop.warehouse.location". If I do that then it works! Interestingly enough it works with the table location set like you suggested and also with the full path like I had. Anyway, I'll update the document accordingly to show at least how it can be done currently. Thanks!

**Comment by massdosage** [comment]
This is now ready for review, I'd appreciate a look from anyone who has time. @pvary @shardulm94 @rdblue @rdsr @marton-bod

File: site/docs/hive.md [hunk 3]
```diff
@@ -79,12 +79,45 @@ In order to query a Hive table created by either of the HiveCatalog methods desc
 ```sql
 SET iceberg.mr.catalog=hive;
 ```
-You should now be able to issue Hive SQL `SELECT` queries using the above table and see the results returned from the underlying Iceberg table. Both the Map Reduce and Tez query execution engines are supported.
```
Comment (pvary) [inline site/docs/hive.md:82]: Why did we remove that both MR and Tez is supported?

File: site/docs/hive.md [hunk 4]
```diff
@@ -79,12 +79,45 @@ In order to query a Hive table created by either of the HiveCatalog methods desc
...
+```sql
+CREATE EXTERNAL TABLE database_a.table_c 
+STORED BY 'org.apache.iceberg.mr.hive.HiveIcebergStorageHandler' 
+LOCATION 'hdfs://some_bucket/path_to_hadoop_warehouse/database_a/table_c'
+TBLPROPERTIES (
+  'iceberg.mr.catalog'='hadoop', 
+  'iceberg.mr.catalog.hadoop.warehouse.location'='hdfs://some_bucket/path_to_hadoop_warehouse')
+;
+```
+Note that the Hive database and table name *must* match the values used in the Iceberg `TableIdentifier` when the table was created.
```
Comment (pvary) [inline site/docs/hive.md:104]: It would be good to remove this restriction later

File: site/docs/hive.md (hunk 3 above)
Comment (massdosage) [reply site/docs/hive.md:82]: > Why did we remove that both MR and Tez is supported?

I moved it to the bottom under "Features" so I didn't have to keep repeating it :)

Comment (marton-bod) [reply site/docs/hive.md:82]: It has been moved down here: https://github.com/apache/iceberg/pull/1837/files#diff-0270f04c6c1a4be5da895415fff2797103da7ded6ec97c303f2f7e218e99ac26R122 to avoid duplication :)

File: site/docs/hive.md [hunk 5]
```diff
@@ -79,12 +79,45 @@ In order to query a Hive table created by either of the HiveCatalog methods desc
...
+You should now be able to issue Hive SQL `SELECT` queries using the above table and see the results returned from the underlying Iceberg table.
 ```sql
 SELECT * from table_b;
 ```
 
+#### Using Hadoop Catalog
+Iceberg tables created using `HadoopCatalog` are stored entirely in a directory in a filesytem like HDFS. 
+
+##### Create an Iceberg table
+The first step is to create an Iceberg table using the Spark/Java/Python API and `HadoopCatalog`. For the purposes of this documentation we will assume that the fully qualified table identifier is `database_a.table_c` and that the Hadoop Catalog warehouse location is `hdfs://some_bucket/path_to_hadoop_warehouse`. Iceberg will therefore create the table at the location `hdfs://some_bucket/path_to_hadoop_warehouse/database_a/table_c`.
```
Comment (pvary) [inline site/docs/hive.md:91]: After the latest changes now it is possible to use Hive `CREATE TABLE` command to create the Hive table and the Iceberg table at the same time. Do we want to document it here, or do we want to do it in a different PR?

File: site/docs/hive.md (hunk 3 above)
Comment (pvary) [reply site/docs/hive.md:82]: Ok guys, just do not pile on me please 😄
I am reading stuff sequentially 😄

File: site/docs/hive.md [hunk 6]
```diff
@@ -79,12 +79,45 @@ In order to query a Hive table created by either of the HiveCatalog methods desc
...
+
+##### Create a Hive table
+Now overlay a Hive table on top of this Iceberg table by issuing Hive DDL like so:
+```sql
+CREATE EXTERNAL TABLE database_a.table_c 
+STORED BY 'org.apache.iceberg.mr.hive.HiveIcebergStorageHandler' 
+LOCATION 'hdfs://some_bucket/path_to_hadoop_warehouse/database_a/table_c'
+TBLPROPERTIES (
+  'iceberg.mr.catalog'='hadoop',
```
Comment (rdblue) [inline site/docs/hive.md:100]: Does it work to set `iceberg.mr.catalog` in table properties for a Hive catalog table as well? If so, we should probably note that a Hive table can be created for an Iceberg table this way as well.

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Thanks, @massdosage! This looks great. I'll merge and redeploy the docs.

File: site/docs/hive.md (hunk 5 above)
Comment (rdblue) [reply site/docs/hive.md:91]: I think we should add this, but in a separate PR because this one is ready to merge.

File: site/docs/hive.md (hunk 6 above)
Comment (massdosage) [reply site/docs/hive.md:100]: I think so, I'm not sure why we don't set that when we automatically create those tables when the Hive exec engine is enabled. I'll make a note and take a look and will update the documentation accordingly (and perhaps also add the code to set that property).

File: site/docs/hive.md (hunk 5 above)
Comment (massdosage) [reply site/docs/hive.md:91]: @pvary - sounds good. What I've done for everything I've documented here is first I've added code to an integration test suite I have that uses the Spark and Hive runtimes on a distributed cluster to really try test everything as an end user would. Can you give me the high level steps needed for what you're talking about above and I'll try reproduce it and can then raise a separate PR to document it? I assume this is related to the first parts of the output format that were merged?

Comment (pvary) [reply site/docs/hive.md:91]: This should work now for HiveCatalog:
```
CREATE EXTERNAL TABLE customers (customer_id BIGINT, first_name STRING, last_name STRING)
STORED BY 'org.apache.iceberg.mr.hive.HiveIcebergStorageHandler'
```
This should work for other Catalogs:
```
CREATE EXTERNAL TABLE customers (customer_id BIGINT, first_name STRING, last_name STRING)
STORED BY 'org.apache.iceberg.mr.hive.HiveIcebergStorageHandler'
LOCATION '<LOCATION>'
```

The location and the table identifier should be the same ATM.
